### PR TITLE
Always pass original input code to the interface

### LINKED ISSUE #224: getCodeSize after callCode
We found this issue while debugging [stCallCodes/callcallcode_01](https://github.com/ethereum/tests/blob/develop/GeneralStateTests/stCallCodes/callcallcode_01.json#L45-L59) with evm2wasm, and the codeSize WIP patch https://github.com/ewasm/hera/pull/217/.

Here's the correct trace, executing cpp-ethereum without hera:
```
test/testeth -t GeneralStateTests/stCallCodes -- --testpath ~/cpp-ethereum/test/jsontests --singlenet "Byzantium" --singletest "callcallcode_01" --jsontrace '{"disableStack": false, "fullStorage": false, "disableStorage": true, "disableMemory": true}'
```
trace output: https://gist.github.com/hugo-dc/487b7f3021f2aaadbcc26dcb2c62307b

Here's the trace with a patched Hera:
```
test/testeth -t GeneralStateTests/stCallCodes -- --testpath ~/cpp-ethereum/test/jsontests --singlenet "Byzantium" --singletest "callcallcode_01" --vm hera --evmc evm2wasm=true --evmc evm2wasm.js=true --evmc evm2wasm.js-trace=true --evmc fallback=false -d 0 
```
trace output: https://gist.github.com/hugo-dc/e7ec65c8d92f103e5c8c49bd7307e7ac

The key lines are around CALLCODE:
```
evmTrace
{"depth":1,"gas":352279,"gasCost":21,"op":CALLCODE,"pc":35,"stack":["0x000000000000000000000000000000000000000000000000000000000003d090","0x0000000000000000000000001000000000000000000000000000000000000002","0x0000000000000000000000000000000000000000000000000000000000000002","0x0000000000000000000000000000000000000000000000000000000000000000","0x0000000000000000000000000000000000000000000000000000000000000040","0x0000000000000000000000000000000000000000000000000000000000000000","0x0000000000000000000000000000000000000000000000000000000000000040"]}
useGas 0
useGas 6
~~~ call_message.destination: 0x1000000000000000000000000000000000000002
~~~ call_message.sender: 0x1000000000000000000000000000000000000001
$callCode 3d090 a0 80 8428 40
Calling evm2wasm.js (input 33 bytes)...
...
Executing...
hera.cpp execute...
~~~ msg.destination: 0x1000000000000000000000000000000000000001
~~~ msg.sender: 0x1000000000000000000000000000000000000001
importGlobals
evmTrace
{"depth":2,"gas":252300,"gasCost":0,"op":PUSH1,"pc":0,"stack":[]}
...
```
And then the CODESIZE inside callCode:
```
evmTrace
{"depth":2,"gas":132269,"gasCost":21,"op":CODESIZE,"pc":25,"stack":[]}
useGas 0
getCodeSize
~~~ getCodeSize|msg.destination: 0x1000000000000000000000000000000000000001
               |msg.sender: 0x1000000000000000000000000000000000000001
~~~ code_size: 33
0x1000000000000000000000000000000000000002
evmTrace
{"depth":2,"gas":132267,"gasCost":21,"op":PUSH1,"pc":26,"stack":["0x0000000000000000000000000000000000000000000000000000000000000021"]}
useGas 3
evmTrace
{"depth":2,"gas":132264,"gasCost":24,"op":SSTORE,"pc":28,"stack":["0x00000000000000000000000000000000000000000000000000000000000000ee","0x0000000000000000000000000000000000000000000000000000000000000021"]}
useGas 0
```

We forced the test to pass by patching getCodeSize: https://github.com/ewasm/hera/blob/8eea45037d4e214ccabb55b7b1af5a08767c2c3d/src/eei.cpp#L370-L381

With this code:
```
if (import->base == Name("getCodeSize")) {
  heraAssert(arguments.size() == 0, string("Argument count mismatch in: ") + import->base.str);

  HERA_DEBUG << "getCodeSize\n";

  takeGas(GasSchedule::base);
  cout << "~~~ getCodeSize|msg.destination: " << addrToHex(msg.destination) << endl;
  cout << "               |msg.sender: " << addrToHex(msg.sender) << endl;
  evmc_address address = msg.destination;
  address.bytes[19] = 2;
  size_t code_size = context->fn_table->get_code_size(context, &address);
  cout << "~~~ code_size: " << code_size << endl;
  cout << addrToHex(address) << endl;
  return Literal(static_cast<uint32_t>(code_size));
}
```

The destination of callCode is `0x100000....00002`, but inside the execution context of callCode this destination address is lost. From within getCodeSize we want to get the code size of `0x100000...0000002`, but both `msg.destination` and `msg.sender` are `0x100000000...000000001`. So we manually set the address passed to `get_code_size()` to `0x100000000...000000002`, and getCodeSize then returns the expected result.

### PR DESCRIPTION
**Pull request by axic**
Fixes #224. Closes #217.

### CODE REVIEW

File: src/hera.cpp [hunk 1]
```diff
@@ -218,6 +218,7 @@ vector<uint8_t> evm2wasm(evmc_context* context, vector<uint8_t> const& input) {
 void execute(
   evmc_context* context,
   vector<uint8_t> const& code,
+  vector<uint8_t> const& state_code,
```
Comment (axic) [inline src/hera.cpp]: Better naming might be needed.

Comment (jakelang) [reply src/hera.cpp]: `input_code` or `initial_code`?

**Comment by cdetrio** [comment]
```
evmTrace
{"depth":1,"gas":604600,"gasCost":21,"op":PUSH1,"pc":36,"stack":["0x0000000000000000000000000000000000000000000000000000000000000001"]}
useGas 3
evmTrace
{"depth":1,"gas":604600,"gasCost":24,"op":SSTORE,"pc":38,"stack":["0x0000000000000000000000000000000000000000000000000000000000000001","0x0000000000000000000000000000000000000000000000000000000000000001"]}
useGas 0
storageStore 20 0
evmTrace
{"depth":0,"gas":3583600,"gasCost":21,"op":PUSH1,"pc":36,"stack":["0x0000000000000000000000000000000000000000000000000000000000000001"]}
useGas 3
evmTrace
{"depth":0,"gas":3583600,"gasCost":24,"op":SSTORE,"pc":38,"stack":["0x0000000000000000000000000000000000000000000000000000000000000000","0x0000000000000000000000000000000000000000000000000000000000000001"]}
useGas 0
storageStore 20 06
/home/ewasm/hugo-dc/cpp-ethereum/test/tools/libtesteth/ImportTest.cpp(257): fatal error: in "GeneralStateTests/stCallCodes": Error when comparing states: preBalance + miningReward < postBalance (3000000000000000000 < 115792089237316195423570985008687907853269984665640564039460584007913129639936) callcallcode_01

*** 1 failure is detected (5 failures are expected) in the test module "Master Test Suite"
```

**Comment by cdetrio** [comment]
the reference trace for cpp-ethereum running `stCallCodes/callcallcode_01` is here: https://gist.github.com/hugo-dc/487b7f3021f2aaadbcc26dcb2c62307b

**Pull request merged**